luacompiler/src/lib/irgen/register_map.rs

```diff
@@ -1,43 +1,80 @@
-use std::{cell::RefCell, collections::HashMap};
+use std::{collections::HashMap, vec::Vec};
 
-/// Represents a structure that is used to keep track of the mapping
-/// between Lua variables and register ids.
-pub struct RegisterMap {
-    reg_count: RefCell<u8>,
-    reg_map: RefCell<HashMap<String, u8>>,
+/// Represents a tuple which is used to specify the lifetime of a register.
+/// For example if a register is first used by the 4th instruction of the bytecode, and
+/// used last by the 7th instruction, the register's lifetime would be (4, 8).
+pub struct Lifetime(usize, usize);
+
+impl Lifetime {
+    pub fn new(sp: usize) -> Lifetime {
+        Lifetime(sp, sp + 1)
+    }
+
+    /// Get the start point of the register.
+    pub fn start_point(&self) -> usize {
+        self.0
+    }
+
+    /// Get the end point of the register.
+    pub fn end_point(&self) -> usize {
+        self.1
+    }
+
+    fn set_end_point(&mut self, ep: usize) {
+        self.1 = ep
+    }
+}
+
+/// Represents a structure that is used to map Lua variables to registers, and to keep
+/// track of their lifetimes.
+pub struct RegisterMap<'a> {
+    current_instr: usize,
+    lifetimes: Vec<Lifetime>,
+    reg_map: HashMap<&'a str, usize>,
 }
 
-impl RegisterMap {
-    pub fn new() -> RegisterMap {
+impl<'a> RegisterMap<'a> {
+    pub fn new() -> RegisterMap<'a> {
         RegisterMap {
-            reg_count: RefCell::new(0),
-            reg_map: RefCell::new(HashMap::new()),
+            current_instr: 0,
+            lifetimes: vec![],
+            reg_map: HashMap::new(),
         }
     }
 
-    /// Generates a fresh register and returns it.
-    /// This is used in cases like `x = 1 + 2 + 3` to generate intermmediate
-    /// registers in which, for instance, the result of 2 + 3 is stored.
-    pub fn new_reg(&self) -> u8 {
-        let to_return = *self.reg_count.borrow();
-        *self.reg_count.borrow_mut() += 1;
-        to_return
+    /// Increments the current instruction by 1. This method is used by the IR generator
+    /// in order to signify that a new instruction has been processed.
```

Comment:
I don't get the point of this function. Why not just increment `current_instr` directly?